File: src/edu/uf/interactable/afumigatus/TAFC.cpp
/*
 * TAFC.cpp
 *
 *  Created on: Jul 19, 2024
 *      Author: henriquedeassis
 */

#include "TAFC.h"
#include "../../utils/Id.h"
#include "../PositionalInfectiousAgent.h"
#include "../../primitives/Interactions.h"
#include "../../utils/Constants.h"
#include "../Siderophore.h"

namespace edu {
namespace uf {
namespace interactable {

using namespace edu::uf::utils;
using namespace edu::uf::primitives;
using namespace edu::uf::diffusion;

const int TAFC::classId = Id::getId();


const string TAFC::NAME = "TAFC";
const double TAFC::THRESHOLD = utils::constexprants::K_M_TF_TAFC * utils::constexprants::VOXEL_VOL / 1.0e6;
TAFC* TAFC::molecule = nullptr;

TAFC::TAFC(double**** qttys, int xbin, int ybin, int zbin, char wbin, Diffuse* diffuse)
    : Siderophore::Siderophore(qttys, xbin, ybin, zbin, wbin, diffuse) {}

TAFC* TAFC::getMolecule(Diffuse* diffuse, int xbin, int ybin, int zbin, char wbin) {
	if (molecule == nullptr) {
	        double**** values = new double***[wbin];
	        for (int i = 0; i < wbin; ++i) {
	            values[i] = new double**[xbin];
	            for (int j = 0; j < xbin; ++j) {
	                values[i][j] = new double*[ybin];
	                for (int k = 0; k < ybin; ++k) {
	                    values[i][j][k] = new double[zbin]();
	                }
	            }
	        }
	        molecule = new TAFC(values, xbin, ybin, zbin, wbin, diffuse);
	    }
	    return molecule;
}

int TAFC::getClassId() const {
	return classId;
}

double TAFC::getKd() {
    return -1;
}

void TAFC::degrade() {}

void TAFC::computeTotalMolecule(int x, int y, int z) {
	totalMoleculesAux[0] += get(0, x, y, z);
	totalMoleculesAux[1] += get(1, x, y, z);
}

bool TAFC::templateInteract(Interactable* interactable, int x, int y, int z) {
    /*if (auto transferrin = dynamic_cast<Transferrin*>(interactable)) {
        return Interactions::siderophoreTransferrinChelation(dynamic_cast<Molecule*>(transferrin), this, Constants::K_M_TF_TAFC, x, y, z);
    }*/
    if (auto afumigatus = dynamic_cast<PositionalInfectiousAgent*>(interactable)) {
        Interactions::uptakeSiderophore(afumigatus, this, utils::constexprants::TAFC_UP, x, y, z);
        return Interactions::secreteSiderophore(afumigatus, this, x, y, z);
    }
    /*if (auto iron = dynamic_cast<Iron*>(interactable)) {
        return Interactions::siderophoreIronChelation(dynamic_cast<Molecule*>(iron), this, x, y, z);
    }*/
    return interactable->interact(this, x, y, z);
}

string TAFC::getName() const {
    return NAME;
}

double TAFC::getThreshold() const {
    return THRESHOLD;
}

int TAFC::getNumState() const {
    return NUM_STATES;
}

double TAFC::getSiderophoreQtty() {
    return utils::constexprants::TAFC_QTTY;
}

} /* namespace interactable */
} /* namespace uf */
} /* namespace edu */

File: java/src/edu/uf/interactable/Afumigatus/TAFC.java
package edu.uf.interactable.Afumigatus;

import edu.uf.Diffusion.Diffuse;
import edu.uf.compartments.GridFactory;
import edu.uf.interactable.Interactable;
import edu.uf.interactable.Iron;
import edu.uf.interactable.Molecule;
import edu.uf.interactable.PositionalInfectiousAgent;
import edu.uf.interactable.Siderophore;
import edu.uf.interactable.Transferrin;
import edu.uf.primitives.Interactions;
import edu.uf.utils.Constants; 

public class TAFC extends Siderophore{
    

	public static final String NAME = "TAFC";
	public static final int NUM_STATES = 2;
    private static final double THRESHOLD = Constants.K_M_TF_TAFC * Constants.VOXEL_VOL / 1.0e6;
    
    private static TAFC molecule = null;
    
    
    private TAFC(double[][][][] qttys, Diffuse diffuse) {
    	super(qttys, diffuse, NAME);
	}
    
    public static TAFC getMolecule(Diffuse diffuse) {
    	if(molecule == null) {
    		double[][][][] values = new double[NUM_STATES][GridFactory.getXbin()][GridFactory.getYbin()][GridFactory.getZbin()];
    		molecule = new TAFC(values, diffuse); 
    	}
    	return molecule;
    }
    
    public static TAFC getMolecule() {
    	return getMolecule(null);
    }
    
    @Override
    public double getKd() {
    	return -1;
    }
    
    public void degrade() {} //REVIEW

    public void computeTotalMolecule(int x, int y, int z) {
    	this.totalMoleculesAux[0] = this.totalMoleculesAux[0] + this.get(0, x, y, z);
    	this.totalMoleculesAux[1] = this.totalMoleculesAux[1] + this.get(0, x, y, z);
    }

    protected boolean templateInteract(Interactable interactable, int x, int y, int z) {
        if (interactable instanceof Transferrin) {
        	//Interactions.transferrinIronChelation((Molecule) interactable, this, x, y, z);
            return Interactions.siderophoreTransferrinChelation((Molecule) interactable, this, Constants.K_M_TF_TAFC, x, y, z);
        }
        if (interactable instanceof Afumigatus) {
        	PositionalInfectiousAgent af = (PositionalInfectiousAgent) interactable;
        	Interactions.uptakeSiderophore(af, this, Constants.TAFC_UP, x, y, z);
            return Interactions.secreteSiderophore(af, this, x, y, z);
        }
        if (interactable instanceof Iron)
        	return Interactions.siderophoreIronChelation((Molecule) interactable, this, x, y, z);

        return interactable.interact(this, x, y, z);
    }

	@Override
	public String getName() {
		return NAME;
	}
 
	@Override
	public double getThreshold() {
		return THRESHOLD;
	}

	@Override
	public int getNumState() {
		return NUM_STATES;
	}
	
	@Override
	public boolean isTime() {
		return true;
	}

	@Override
	public double getSiderophoreQtty() {
		return Constants.TAFC_QTTY;
	}
}
